app/multitenant/dynamo_collector.go

```diff
@@ -240,37 +240,61 @@ func (c *dynamoDBCollector) getCached(reportKeys []string) ([]report.Report, []s
 	return foundReports, missingReports
 }
 
+// Fetch multiple reports in parallel from S3.
 func (c *dynamoDBCollector) getNonCached(reportKeys []string) ([]report.Report, error) {
-	reports := []report.Report{}
+	type result struct {
+		key    string
+		report *report.Report
+		err    error
+	}
+
+	ch := make(chan result, len(reportKeys))
+
 	for _, reportKey := range reportKeys {
-		var resp *s3.GetObjectOutput
-		err := timeRequest("Get", s3RequestDuration, func() error {
-			var err error
-			resp, err = c.s3.GetObject(&s3.GetObjectInput{
-				Bucket: aws.String(c.bucketName),
-				Key:    aws.String(reportKey),
-			})
-			return err
-		})
-		if err != nil {
-			return nil, err
-		}
-		reader, err := gzip.NewReader(resp.Body)
-		if err != nil {
-			log.Errorf("Error gunzipping report: %v", err)
-			continue
-		}
-		rep := report.MakeReport()
-		if err := codec.NewDecoder(reader, &codec.MsgpackHandle{}).Decode(&rep); err != nil {
-			log.Errorf("Failed to decode report: %v", err)
-			continue
+		go func(reportKey string) {
+			r := result{key: reportKey}
+			r.report, r.err = c.getNonCachedReport(reportKey)
+			ch <- r
+		}(reportKey)
+	}
+
+	reports := []report.Report{}
+	for range reportKeys {
+		r := <-ch
+		if r.err != nil {
+			return nil, r.err
```

Comment:
I think this will leave orphaned goroutines if the first response contains an error.